### PR TITLE
Price Endpoint: Construct Trade object using external price information

### LINKED ISSUE #534: Price Update
**Is your feature request related to a problem? Please describe.**
I encountered a price update for the first time.It came up quite unexpected and I was wondering what the prior price was. Might be cool to display the old price or more prominent if it is already displayed

### PR DESCRIPTION
**Pull request by W3stside**
Closes #534 

Reworks the creation of our `Trade` object inside `swap/hooks` to consume the price endpoint data

### CODE REVIEW

**Review by anxolin — COMMENTED** [review]
I had an issue with the PR, but this PR uses the wrong API (see #612)

So it makes the numbers to be off (so, while this is not solved, this has to be tested locally, cc: @MareenG u might want to hold the test a bit until we have it ready for testing)
image
more context: https://gnosisinc.slack.com/archives/CPZA1AGKY/p1620640554004000

One small issue I notice, is that I cannot get exactly the same numbers by computing what the endpoints return. Not sure if I do something wrong or we have a calculation issue:
image

@W3stside are we converting to floats at any point or something that might be breaking up the calculations?

**Comment by anxolin** [comment]
## instructions for QA
> For now we can hold this tests until the PR is testable

This PR displays **our own price** instead of Uniswap price.

Our own price, at the time of the writing is the best between 1inch, uniswap, sushiswap

**Test instructions:**
- Verify that given the displayed values for PRICE and FEE, the minimum received, maximun sold, and the FROMs and TOs are correct for both the buy and sell orders
- Prices make sense, if we check in these exchanges, we have the same or better price
- The price and fee should update automatically

**Techincal test instructions** (technical tests)
- Make sure that the Fee and Price is queries and updated
- Make sure the displayed values and the API endpoint matches

**Comment by W3stside** [comment]
> One small issue I notice, is that I cannot get exactly the same numbers by computing what the endpoints return. Not sure if I do something wrong or we have a calculation issue:
> image
> 
> @W3stside are we converting to floats at any point or something that might be breaking up the calculations?

the amount returned from the price endpoint is parameterised using pre-fee amounts, meaning for a buy order the calculation you're doing is wrong because you're passing a pre-fee inputAmount, when it should be `(priceAmountFromEndpoint + feeInAtoms) / 10**18 * (1 + 0.005)`

File: src/custom/state/swap/extension.ts [hunk 1]
```diff
@@ -85,72 +51,102 @@ export const stringToCurrency = (amount: string, currency: Currency) =>
...
-  return extendExactInTrade({
-    exactInTrade: inTrade,
-    originalTrade,
-    typedAmountAsCurrency: parsedAmount,
-    fee
+  // make sure we have a typed in amount, a fee, and a price
+  // else we can assume the trade will be null
+  if (!parsedAmount || !originalTrade || !outputCurrency || !quote?.fee || !quote?.price) return null
+
+  const feeAsCurrency = stringToCurrency(quote.fee.amount, parsedAmount.currency)
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/custom/state/swap/extension.ts:63]: EDIT: I see is correct, cause this is a sellOrder

We have input and output currency, Doesn't it make sense to use that instead the `inputAmount`? (not sure if it's equivalent). My reason to say this, is that if the fee is always in the FROM, and we have it in `inputAmount` maybe be is safer than using `parseAmount` that I understand may depend if we are buy or sell order.

Not sure im right or terribly wrong. just thinking aloud

**Review by anxolin — COMMENTED** [review]
Really good job! 

I'm only having some concerns with the calculation, because I don't get exactly your same numbers as I mentioned in the last comment (small decimal differences). Other than that, looks good to me

File: src/custom/state/swap/extension.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/state/swap/extension.ts:63]: i can make it more explicit by passing the more relevantly named amount

File: src/custom/state/swap/extension.ts [hunk 2]
```diff
@@ -10,69 +11,34 @@ export type TradeWithFee = Trade & {
...
-export function extendExactOutTrade(params: Omit<ExtendedTradeParams, 'exactInTrade'>): TradeWithFee | null {
-  const { exactOutTrade, fee } = params
-
-  if (!exactOutTrade || !fee.feeAsCurrency) return null
+function _constructTradePrice({
+  sellToken,
+  buyToken,
+  kind,
+  price
+}: CanonicalMarketParams<CurrencyAmount | undefined> & { price?: PriceInformation }) {
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/state/swap/extension.ts:19]: ok i will restructure this

**Comment by W3stside** [comment]
> image
> 
> @W3stside are we converting to floats at any point or something that might be breaking up the calculations?

I think it is correct, see the `maximumAmountOut` method from the `Trade` object:
```ts
function maximumAmountIn(slippageTolerance) {
    !!slippageTolerance.lessThan(ZERO) ? process.env.NODE_ENV !== "production" ? invariant(false, 'SLIPPAGE_TOLERANCE') : invariant(false) : void 0;

    if (this.tradeType === TradeType.EXACT_INPUT) {
      return this.inputAmount;
    } else {
      var slippageAdjustedAmountIn = new Fraction(ONE).add(slippageTolerance).multiply(this.inputAmount.raw).quotient;
      return this.inputAmount instanceof TokenAmount ? new TokenAmount(this.inputAmount.token, slippageAdjustedAmountIn) : CurrencyAmount.ether(slippageAdjustedAmountIn);
    }
  }
```

I logged each param to recreate this calculation:
Screenshot 2021-05-10 at 12 14 26

**Comment by anxolin** [comment]
@W3stside @alfetopito I documented the issue I'm having with the calculation. Maybe double check I'm doing correctly the math https://docs.google.com/document/d/1zxY_9LH7yPKwXbpwQnN5cGkKWYh8j-Fnl9ovCCI4pKs/edit

**Comment by anxolin** [comment]
This PR is reviewable @alfetopito @MareenG @michelbio let's put some eyes on it

**Comment by W3stside** [comment]
@anxolin @alfetopito @MareenG I've made changes as necessary using the latest changes from Anxo's #620 and checking calculations. Slippage is no longer applied to the `fee` as it was doing previously. 

For **buy** orders, formula is now:
inputWithSlippage = `inputAmountWithoutFees * (1 + slippage) + feeAmount`

sell orders slippage calc is the same. this changes as the trade sdk from uni was calculating all this internally using `this.inputAmount` which needed to be switched in this PR for buy orders to be the input amount without fees applied.

refer to logs for correct amounts

**Review by anxolin — COMMENTED** [review]
Still doesn't work for me the numbers. 
There's a small difference in the minimum received.

Details in https://docs.google.com/document/d/1zxY_9LH7yPKwXbpwQnN5cGkKWYh8j-Fnl9ovCCI4pKs/edit#


It would be great if we could unit test or something, so we can reduce the errors and make sure our calculations are accurate

**Comment by W3stside** [comment]
@anxolin waterfalls are in here

tests here: https://github.com/gnosis/gp-swap-ui/pull/611/files#diff-db9619891a480a5d599cce4e7086f60863b6b8c442d3d70b5b02c03677a4fb29

**Comment by W3stside** [comment]
merging in to release 0.12 @anxolin  - can test there and open into it to fix

**Pull request merged**